common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -180,7 +180,10 @@ namespace com.keyman.text {
       //        once THOSE are properly relocated.  (They're too DOM-heavy to remain in web-core.)
 
       // Only return true (for the eventual event handler's return value) if we didn't match a rule.
-      return ruleBehavior;
+
+      // Alas, the above notes are the goal, and will be separately addressed by #3695.  For now, a hacky
+      // workaround in order to not break default behaviors.
+      return ruleBehavior && !ruleBehavior.triggerKeyDefault;  // return `true` if the key is considered fully handled.
```

Comment:
I suspect this workaround is resulting in the intermittent behavior:
1. <kbd>TAB</kbd> from the external keyboard
2. KMW generates `\t` and this returns true
3. KMEA gets the status (true), and thinks KMW didn't process the key, so KMEA dispatches another Tab key to get handled.

keyboard.html (ll. 331-334)
```
         if (kmw['executeHardwareKeystroke'](code, shift, lstates)) { // false if matched, true if not
          // KMW didn't process the key, so have the Android app dispatch the key with the original event modifiers
          window.jsInterface.dispatchKey(code, eventModifiers);
        }
```

Reply:
I added a console.warn and can confirm with the Galaxie Hebrew keyboard
<kbd>m</kbd><kbd>TAB</kbd> generates the transformed character along with an extra `\t`.

Reply:
> I suspect this workaround is resulting in the intermittent behavior:
> 
>     1. TAB from the external keyboard
> 
>     2. KMW generates `\t` and this returns true
> 
>     3. KMEA gets the status (true), and thinks KMW didn't process the key, so KMEA dispatches another Tab key to get handled.
> 
> 
> keyboard.html (ll. 331-334)
> 
> ```
>          if (kmw['executeHardwareKeystroke'](code, shift, lstates)) { // false if matched, true if not
>           // KMW didn't process the key, so have the Android app dispatch the key with the original event modifiers
>           window.jsInterface.dispatchKey(code, eventModifiers);
>         }
> ```

That sounds like what I expected with this:

> How frequently do you see the 'doubling' effect? Also, is there any form of regularity you can see? (Such as... the key immediately after a TAB is always doubled?)

KMW hadn't been processing it properly but now is... but there's then the matter of how to pass this along to Android.  Especially for cases where a command-style TAB (for change of focus) is expected.

Reply:
I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.

Android would then interpret Tab depending on the field.

Reply:
> I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.
> 
> Android would then interpret Tab depending on the field.

Just to be clear... just for hardware-based keystrokes, right?  OSK-based keystrokes require KMW emulation/handling, after all.  Or is there a way to make an OSK-based keystroke _act_ like a hardware keystroke on Android?

Reply:
So for a keyboard like galaxie_hebrew_positional, KMW should match K_TAB and return `false` to KMEA.
For sil_euro_latin, KMW doesn't match K_TAB and should return `true` to KMEA to handle the tab.

Reply:
> I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.
> 
> Android would then interpret Tab depending on the field.

> So for a keyboard like galaxie_hebrew_positional, KMW should match K_TAB and return `false` to KMEA.

Pick one, please.  The first quote above indicates that I should return `true` to KMEA.  Hence my question above:




> > I think from #3664 we said for KMW to **not** process K_TAB or K_ENTER and let the embedded devices handle it.
> > Android would then interpret Tab depending on the field.
> 
> Just to be clear... just for hardware-based keystrokes, right? OSK-based keystrokes require KMW emulation/handling, after all. Or is there a way to make an OSK-based keystroke _act_ like a hardware keystroke on Android?

Reply:
@marc - My interpretation of all of this:

For keystrokes that match _a_ rule, but not within the final processing group:

- When embedded and the keystroke was received from a physical keyboard:
    - KMW should NOT emit the keystroke's default behavior
    - KMW should indicate that it was unhandled, thus allowing KMEA to do it.
    - KMEA should then update the text-context accordingly, allowing predictive text to fill it in.
- When the keystroke was received from the OSK:
    - KMW should emit the keystroke's default behavior.
        - Note:  when in 'native', within a browser this can trigger focus changes.
        - Is that _possible_ to do from 'synthetic' keystrokes for Android?  Or do we just give up there?
- When not embedded and the keystroke was received from a physical keyboard:
    - KMW should not (directly) emit the keystroke's default behavior
    - The key-handling event handlers indicate that the event wasn't handled, triggering the browser to do perform default behaviors
    - Predictive text, if active, should note any context changes triggered as a result (NOTE:  not currently done!) and update predictions accordingly.

My issues with this breakdown so far:
- I don't think this will (currently) be handled well by Android default-handling rules.
    - What if a plain-text key had a similar setup?
    - Would our Android app properly emit that 'a'?  My impression is that it has special handling for `K_TAB` and `K_ENTER`, but I could be wrong.
- This markedly complicates predictive text.
    - For `galaxie_hebrew_positional`, this behavior only exists for `K_TAB` and `K_ENTER`... so we kind of luck out re: predictive text.
    - Suppose a key's default behavior is to emit 'a' and has a matching rule (but not in the final match group)... and wasn't pressed.  How would predictive text know to include that 'a'?
    - My current approach:  for possible 'alternate' keystrokes, follow 'synthetic'/OSK-based rules:  always model text-based default behaviors (except for TAB and ENTER in 'native' mode)
    - The issue is then that the original keystroke (from hardware) doesn't come from the same processing path as the fat-finger alternates (from synthetic / OSK processing).
    - There's _also_ the issue of recognizing _after the fact_ that the emitted default behavior (example: 'a') is actually a part of the previously-processed keystroke, rather than treating it as a new, separate 'keystroke' upon the (delayed) context update.

Reply:
For @darcywong00:

> Android would then interpret Tab depending on the field.

_When_ does Android's OS allow us to do this?  Can we do this at any time with the correct setup?

_What_ is required, by Android's APIs, to inform the Android OS that this should be done?